app/coffee/ResourceWriter.coffee

```diff
@@ -5,13 +5,53 @@ async = require "async"
 mkdirp = require "mkdirp"
 OutputFileFinder = require "./OutputFileFinder"
 Metrics = require "./Metrics"
+Errors = require "./Errors"
 logger = require "logger-sharelatex"
 settings = require("settings-sharelatex")
 
 parallelFileDownloads = settings.parallelFileDownloads or 1
 
 module.exports = ResourceWriter =
-	syncResourcesToDisk: (project_id, resources, basePath, callback = (error) ->) ->
+
+	syncResourcesToDisk: (request, basePath, callback = (error) ->) ->
+		if request.syncType is "incremental"
+			ResourceWriter.checkSyncState request.syncState, basePath, (error, syncStateOk) ->
+				logger.log syncState: request.syncState, result:syncStateOk, "checked state on incremental request"
+				return callback new Errors.FilesOutOfSyncError("invalid state for incremental update") if not syncStateOk
+				ResourceWriter.saveIncrementalResourcesToDisk request.project_id, request.resources, basePath, callback
+		else
+			@saveAllResourcesToDisk request.project_id, request.resources, basePath, (error) ->
+				return callback(error) if error?
+				ResourceWriter.storeSyncState request.syncState, basePath, callback
+
+	storeSyncState: (state, basePath, callback) ->
+		stateFile = Path.join(basePath, ".resource-sync-state")
+		if not state? # remove the  file if no state passed in
+			logger.log state:state, basePath:basePath, "clearing sync state"
+			fs.unlink stateFile, (err) ->
+				if err? and err.code isnt 'ENOENT'
+					return callback(err)
+				else
+					return callback()
+		else
+			logger.log state:state, basePath:basePath, "writing sync state"
+			fs.writeFile stateFile, state, {encoding: 'ascii'}, callback
+
+	checkSyncState: (state, basePath, callback) ->
+		stateFile = Path.join(basePath, ".resource-sync-state")
+		fs.readFile stateFile, {encoding:'ascii'}, (err, oldState) ->
+			# ignore errors, return true if state matches, false otherwise (including errors)
```

Comment:
Is this a simple way of ignoring errors when the file doesn't exist? If so, can we make this explicit, but still error on things we wouldn't expect, e.g. permissions, failed reads, etc.